IdentityCore/src/workplacejoin/MSIDWorkPlaceJoinUtilBase.m

```diff
@@ -173,4 +185,90 @@ For asymmetric keys this holds the public key hash which allows digital identity
     return keyPair;
 }
 
++ (MSIDWPJKeyPairWithCert *)getWPJKeysWithTenantId:(__unused NSString *)tenantId context:(__unused id<MSIDRequestContext>)context
+{
+    NSString *teamId = [[MSIDKeychainUtil sharedInstance] teamId];
+    
+    if (!teamId)
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, context, @"Encountered an error when reading teamID from keychain.");
+        return nil;
+    }
+    
+    NSString *legacySharedAccessGroup = [NSString stringWithFormat:@"%@.com.microsoft.workplacejoin", teamId];
+    NSData *tagData = [kMSIDPrivateKeyIdentifier dataUsingEncoding:NSUTF8StringEncoding];
+    // Legacy registrations would have be done using RSA, passing keyType = RSA in query
+    NSDictionary *extraPrivateKeyAttributes = @{ (__bridge id)kSecAttrApplicationTag: tagData,
+                                                 (__bridge id)kSecAttrAccessGroup : legacySharedAccessGroup,
+                                                 (__bridge id)kSecAttrKeyType : (__bridge id)kSecAttrKeyTypeRSA };
+    NSDictionary *extraCertAttributes = @{ (__bridge id)kSecAttrAccessGroup : legacySharedAccessGroup };
+    
+    MSID_LOG_WITH_CTX(MSIDLogLevelInfo, context, @"Checking Legacy keychain for registration.");
+    MSIDWPJKeyPairWithCert *legacyKeys = [self findWPJRegistrationInfoWithAdditionalPrivateKeyAttributes:extraPrivateKeyAttributes certAttributes:extraCertAttributes context:context];
+        
+    if (legacyKeys)
+    {
+        if ([NSString msidIsStringNilOrBlank:tenantId])
+        {
+            // ESTS didn't request a specific tenant, just return default one
+            return legacyKeys;
+        }
+        
+        // Read tenantId for legacy identity
+        NSError *tenantIdError = nil;
+        NSString *registrationTenantId = [MSIDWorkPlaceJoinUtil getWPJStringDataForIdentifier:kMSIDTenantKeyIdentifier context:context error:&tenantIdError];
+        
+        // There's no tenantId on the registration, or it mismatches what server requested, keep looking for a better match. Otherwise, return the identity already.
+        if (!tenantIdError
+            && registrationTenantId
+            && [registrationTenantId isEqualToString:tenantId])
+        {
+            return legacyKeys;
+        }
+    }
+    
+    if (!tenantId)
+    {
+        // default registration should have a tenantId associated.
```

Comment:
I thought this is a legacy registration that doesn't have a tenantId, no?

Reply:
This method will check for both. After this line we are checking for default registration which should have a tenantId , if there is none I see no point in querying for it as there should be a tenantId to query default reg with. Returning early here.

Reply:
Got it, that's a server's tenantId, not client. I was confused